examples/basic/main.go

```diff
@@ -40,7 +42,7 @@ func main() {
 	})
 
 	// Setup a handler to be triggered when a message is received
-	client.HandleMessage(func(m messenger.Message, r *messenger.Response) {
+	messenger.Handlers.HandleMessage(func(ctx context.Context, m messenger.Message, r *messenger.Response) {
```

Comment:
Could you explain the rationalization behind moving from the `messenger.New(...)` model to the `context.Context` model? Is there a particular reason we need to break compatibility on this?

Reply:
In my message handler I need to access request scoped clients that are initialized. 

Heavily truncated example from my code base:

```
func main() {
	// Setup server
	s := newServer()

	// Setup messenger handlers
	messenger.Handlers.HandleMessage(s.messageHandler)

	// Setup router
	http.HandleFunc("/webhook/messenger", s.addEnv(s.messengerHandler))

	// Starts server to receive requests
	appengine.Main()
}

func (s *server) messengerHandler(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	http := ctxHTTP(ctx)

	m := messenger.New(messenger.Options{
		// AppEngine requires urlfetch. Set it in 3rd party lib
		Client:      http,
		Verify:      s.config.messenger.Verify,
		AppSecret:   s.config.messenger.AppSecret,
		VerifyToken: s.config.messenger.VerifyToken,
		Token:       s.config.messenger.Token,
	})

	// Pass incoming messages off to the messenger lib handler
	m.Handle(w, r)
}

func (s *server) newEnv(ctx context.Context) *env {
	fs := newFirestoreClient(ctx, s.config.google.ProjectID)

	e := &env{
		fs:   fs,
		http: urlfetch.Client(ctx),
		bot: bot.New(bot.Options{
			DB: fs.client,
		}),
	}

	if e.fs == nil {
		log.Criticalf(ctx, "firestore client nil")
	}

	if e.http == nil {
		log.Criticalf(ctx, "http urlfetch client nil")
	}

	if e.bot == nil {
		log.Criticalf(ctx, "bot client nil")
	}

	return e
}

func (s *server) addEnv(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx := appengine.NewContext(r)


		env := s.newEnv(ctx)
		ctx = withEnv(ctx, env)
		next(w, r.WithContext(ctx))
	}
}


func (s *server) messageHandler(ctx context.Context, m messenger.Message, r *messenger.Response) {
	b := ctxBot(ctx)

	// save conversation state
	if err = b.Save(ctx, c); err != nil {
		log.Errorf(ctx, "error saving conversation", err)
	}
}
```

My message handler needs to access my bot which *per request* is initialized with a context that is request scoped so connections can be terminated across request boundaries if there is an error, or handle other things like timeouts (and communicating it to the entire request boundary).

The only alternative I can think of is utilizing closures but it felt complicated compared to this solution. 

`ctx` also seems to be a pretty common idiom amongst modern libs (especially google cloud ones) 

Examples:

1. https://godoc.org/cloud.google.com/go/bigquery
2. https://godoc.org/cloud.google.com/go/firestore
3. https://godoc.org/cloud.google.com/go/storage

Reply:
If you use Keybase feel free to message me if needed: https://keybase.io/dfischer